### PR TITLE
Add GitHub scm source implementation for status checks properties

### PR DESCRIPTION
**Pull request by XiongKezhi**
- [x] Implementing https://github.com/jenkinsci/checks-api-plugin/pull/31 to enable customized status checks for GitHub SCM Source projects.

### CODE REVIEW

**Comment by XiongKezhi** [comment]
@timja I meet a problem that my trait (or behavior) cannot persist to a project: I add this `Status Checks Properties` behavior to a project this time, but next time when I open the project configuration again, this behavior disappears. I don't know if I miss anything about implementing a trait?

截屏2020-10-16 上午12 15 56

File: pom.xml [hunk 1]
```diff
@@ -25,7 +25,7 @@
     <useBeta>true</useBeta>
 
     <!-- Jenkins Plug-in Dependencies Versions -->
-    <checks-api.version>1.0.2</checks-api.version>
+    <checks-api.version>2.0.0-SNAPSHOT</checks-api.version>
```
Comment (uhafner) [inline pom.xml:28]: It would make sense to use the associated incrementals build:
https://repo.jenkins-ci.org/incrementals/io/jenkins/plugins/checks-api/

File: src/main/java/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksProperties.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @VisibleForTesting
+    GitHubSCMSourceStatusChecksProperties(final SCMFacade scmFacade) {
+        this.scmFacade = scmFacade;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksProperties.java:33]: Such comments can be removed since the IDE is smart enough

File: src/main/java/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksProperties.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        return source.filter(s -> getStatusChecksTrait(s).isSkip()).isPresent();
+    }
+
+    private GitHubSCMSourceStatusChecksTrait getStatusChecksTrait(final GitHubSCMSource source) {
+        return source.getTraits()
+                .stream()
+                .filter(t -> t instanceof GitHubSCMSourceStatusChecksTrait)
+                .findFirst()
+                .map(t -> (GitHubSCMSourceStatusChecksTrait)t)
+                .orElseThrow(IllegalStateException::new);
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksProperties.java:67]: Are you planning to add a test? It is hard to follow the setup here. 
If there is no trait defined then an exception will be thrown, is this intended?

**Review by uhafner — COMMENTED** [review]
Does this approach work the plain GitSCM as well?

File: src/main/java/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksProperties.java (hunk 3 above)
Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksProperties.java]: No, just a problem in my logic. Fixed.

File: src/main/java/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksProperties.java (hunk 2 above)
Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksProperties.java]: I added more meaningful docs.

**Comment by XiongKezhi** [comment]
> Does this approach work the plain GitSCM as well?

No, I'll implement a trait for GitSCM after this. The logic is mostly the same.

File: src/main/resources/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksTrait/config.jelly [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+
+  <f:entry title="Skip Publishing Status Checks" field="skip">
+    <f:checkbox default="false"/>
+  </f:entry>
+
+  <f:entry title="Status Checks Name" field="name">
```
Comment (timja) [inline src/main/resources/io/jenkins/plugins/checks/github/GitHubSCMSourceStatusChecksTrait/config.jelly:8]:
```suggestion
  <f:entry title="Status checks name" field="name">
```

**Review by timja — APPROVED** [review]
could you add a screenshot showing the UI for this?

**Comment by timja** [comment]
Any docs needed?

**Comment by XiongKezhi** [comment]
> Any docs needed?

ok, I will add it.

File: README.md [hunk 5]
```diff
@@ -23,6 +23,10 @@ This plugin has been installed, alone with the [General API Plugin](https://gith
 
 By listening to the Jenkins builds, this plugin will automatically publish statuses (pending, in progress, and completed) to GitHub.
 
+The status checks can be customized by configuring the "Status Checks Properties" behaviour for you GiHub Source (similar behaviour for Git SCM will be provided soon):
```
Comment (timja) [inline README.md:26]:
```suggestion
The status checks can be customized by configuring the "Status Checks Properties" behaviour for you GitHub Source (similar behaviour for Git SCM will be provided soon):
```

File: README.md [hunk 6]
```diff
@@ -23,6 +23,10 @@ This plugin has been installed, alone with the [General API Plugin](https://gith
...
+
+![Status Checks Properties](docs/images/status-checks-properties.png)
```
Comment (timja) [inline README.md:28]: might want to fix your spelling of customized in the screenshot

File: README.md [hunk 7]
```diff
@@ -21,9 +21,10 @@ This plugin has been installed, alone with the [General API Plugin](https://gith
 
 ![GitHub Status](docs/images/github-status.png)
 
-By listening to the Jenkins builds, this plugin will automatically publish statuses (pending, in progress, and completed) to GitHub.
+This plugin implements [the status checks feature from Checks API Plugin](https://github.com/jenkinsci/checks-api-plugin#build-status-check) to publish statuses (pending, in progress, and completed) to GitHub.
 
-The status checks can be customized by configuring the "Status Checks Properties" behaviour for you GiHub Source (similar behaviour for Git SCM will be provided soon):
+However, this is not a default feature.
```
Comment (timja) [inline README.md:26]: hmm don't we want it to be the default (like it currently is) by just installing this plugin?

Comment (XiongKezhi) [reply README.md:26]: oh, I mixed up, so let me make it clear:

1. in checks-api, we set the default value of `skip` to be true, so users won't get logs when no implementation is installed;
2. in github-checks, we set the default value of `skip` to be false, so users can get statuses as long as they installed the implementation, even they didn't add the trait at all?

**Review by timja — APPROVED** [review]
